src/Dfc.CourseDirectory.Core/Models/Postcode.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Text.RegularExpressions;
+
+namespace Dfc.CourseDirectory.Core.Models
+{
+    public sealed class Postcode : IEquatable<Postcode>
+    {
+        private static readonly Regex _ukPostcodePattern = new Regex(
+            @"^((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\s?[0-9][A-Za-z]{2})$",
+            RegexOptions.Compiled);
+
+        private readonly string _value;
+
+        public Postcode(string value)
+            : this(value, validated: false)
+        {
+        }
+
+        private Postcode(string value, bool validated)
+        {
+            if (!validated)
+            {
+                if (!ValidateAndNormalize(value, out value))
+                {
+                    throw new ArgumentException("Input is not a valid UK postcode.", nameof(value));
```

Comment:
Use `nameof(value)` instead of `"Input"` in exception message.

Reply:
Parameter name is already part of the exception. The message mirrors `int.Parse()` and `Guid.Parse()`.

Reply:
`int.Parse()` and `Guid.Parse()` both throw `FormatException` for invalid input data, neither of which use the ctor with an overload for `paramName`. In the case of int.Parse, the parameter name is called `s` and the message refers to "string input", for `Guid.Parse()` the parameter is called `input` and the exception message refers to "input". From what I can see this doesn't mirror their implementations at all.

The variable you're using and what this exception refers to is called `value`, the phrase "Input" means nothing in the context of this execution.

Reply:
Modified to use `FormatException`. Given that the only public API that throws this exception takes a single input parameter I am happy the message is clear.